src/core/filesystem.cc

```diff
@@ -205,11 +212,327 @@ LocalFileSystem::WriteTextFile(
+    std::string name = object_metadata->name();
+    size_t item_start = name.find(full_dir) + full_dir.size();
+    size_t item_end = name.find(
+        "/", item_start);  // GCS response prepends parent directory name
 
+    // Let set take care of subdirectory contents
+    std::string item = name.substr(item_start, item_end - item_start);
+    contents->insert(item);
+  }
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::GetDirectorySubdirs(
+    const std::string& path, std::set<std::string>* subdirs)
+{
+  RETURN_IF_ERROR(GetDirectoryContents(path, subdirs));
+
+  // Erase non-directory entries...
+  for (auto iter = subdirs->begin(); iter != subdirs->end();) {
+    bool is_dir;
+    RETURN_IF_ERROR(IsDirectory(JoinPath({path, *iter}), &is_dir));
+    if (!is_dir) {
+      iter = subdirs->erase(iter);
+    } else {
+      ++iter;
+    }
+  }
+
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::GetDirectoryFiles(
+    const std::string& path, std::set<std::string>* files)
+{
+  RETURN_IF_ERROR(GetDirectoryContents(path, files));
+
+  // Erase directory entries...
+  for (auto iter = files->begin(); iter != files->end();) {
+    bool is_dir;
+    RETURN_IF_ERROR(IsDirectory(JoinPath({path, *iter}), &is_dir));
+    if (is_dir) {
+      iter = files->erase(iter);
+    } else {
+      ++iter;
+    }
+  }
+
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::ReadTextFile(const std::string& path, std::string* contents)
+{
+  bool exists;
+  RETURN_IF_ERROR(FileExists(path, &exists));
+
+  if (!exists) {
+    return Status(
+        RequestStatusCode::INTERNAL, "File does not exist at " + path);
+  }
+
+  std::string bucket, object;
+  ParsePath(path, &bucket, &object);
+
+  gcs::ObjectReadStream stream = client_->ReadObject(bucket, object);
+
+  if (!stream) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "Failed to open object read stream for " + path);
+  }
+
+  std::string data = "";
+  char c;
+  while (stream.get(c)) {
+    data += c;
+  }
+
+  *contents = data;
+
+  return Status::Success;
+}
+
+Status
+GCSFileSystem::WriteTextFile(
+    const std::string& path, const std::string& contents)
+{
+  std::string bucket, object;
+  ParsePath(path, &bucket, &object);
+
+  gcs::ObjectWriteStream stream = client_->WriteObject(bucket, object);
+
+  if (!stream) {
+    return Status(
+        RequestStatusCode::INTERNAL,
+        "failed to open object write stream for " + path);
+  }
+
+  stream.Close();
+
+  google::cloud::StatusOr<gcs::ObjectMetadata> metadata =
+      std::move(stream).metadata();
+
+  if (!metadata) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Error writing to object at " + path);
+  }
+
+  return Status::Success;
+}
+
+#endif  // TRTIS_ENABLE_GCS
+
+Status
+GetFileSystem(const std::string& path, FileSystem** file_system)
+{
+#ifdef TRTIS_ENABLE_GCS
+  // Check if this is a GCS path (gs://$BUCKET_NAME)
+  if (!path.empty() && !path.rfind("gs://", 0)) {
```

Comment:
Here you need to still detect gs:// but if TRTIS_ENABLE_GCS is false then raise an appropriate error about trtis not being compiled with support for gcs.